blocks/algorithms/__init__.py

```diff
@@ -575,6 +577,64 @@ def compute_steps(self, previous_steps):
         return steps, []
 
 
+class VariableClipping(StepRule):
+    """Clip the maximum norm of individual variables along certain axes.
+
+    This :class:`StepRule` can be used to implement L2 norm constraints on
+    individual hidden units, groups of hidden units, convolutional
+    filters or entire weight tensors. Combine with :class:`Restrict`
+    (and possibly :class:`CompositeRule`), to apply such constraints only
+    to certain variables and/or apply different norm constraints to
+    different variables.
+
+    Parameters
+    ----------
+    threshold : float
+        Maximum norm for a given (portion of a) tensor.
+    axes : iterable, optional
+        An iterable collection of integer axes over which to
+        sum in order to calculate the L2 norm.
+
+    Notes
+    -----
+    Investigations into max-norm regularization date from [SREBRO2005]_.
+    The first appearance of this technique as a regularization method
+    for the weight vectors of individual hidden units in feed-forward
+    neural networks may be [Hinton2012]_.
+
+    .. [SREBRO2005] Nathan Srebro and Adi Shraibman.
+       "Rank, Trace-Norm and Max-Norm". *18th Annual Conference
+       on Learning Theory (COLT)*, June 2005.
+
+    .. [Hinton2012] Geoffrey E. Hinton, Nitish Srivastava,
+       Alex Krizhevsky, Ilya Sutskever, Ruslan R. Salakhutdinov.
+       "Improving neural networks by preventing co-adaptation of
+       feature detectors". arXiv:1207.0580.
+
+    """
+    def __init__(self, threshold, axes=None):
+        axes = axes if axes is not None else ()
+        self.axes = set(axes)
+        self.threshold = shared_floatx(threshold)
+        if len(axes) != len(self.axes):
+            raise ValueError("axes must be unique")
+
+    def compute_step(self, param, previous_step):
+        if any(axis >= previous_step.ndim for axis in self.axes):
+            raise ValueError("Invalid axes {} for {}, ndim={}".format(
+                self.axes, param, previous_step.ndim))
+        squares = tensor.sqr(previous_step)
+        if len(self.axes) == 0:
+            norms = l2_norm([previous_step])
+        else:
+            norms = tensor.sqrt(
+                reduce(lambda t, a: t.sum(axis=a, keepdims=True),
+                       sorted(self.axes), squares))
+        return (previous_step * tensor.switch(norms > self.threshold,
```

Comment:
You were right, of course. I was thinking of the return OrderedDict as a list of Theano updates, when it is in fact not that. Fixed in  83700c8, PTAL.